### ISSUE TITLE
Why the linux bzImage built by buildroot which i made can not run successfully on seL4?

### ISSUE THREAD

**Issue by jason416** [SEL4PROJ/camkes-vm-linux#1]
How to compile Linux running on seL4? i simply use buildroot tool to build a kernel 4.8.16, the same as what you offered, but can't run successfully on seL4 whether in qemu or physic machine. the information is like this below:
--------------------------------------------------------------------------------------------------------
[    0.000000] Linux version 4.8.16 (jason@local) (gcc version 6.3.0 (Buildroot 2017.05) ) #1 SMP Wed Dec 19 21:16:18 CST 2018
[    0.000000] Disabled fast string operations
[    0.000000] x86/fpu: Legacy x87 FPU detected.
[    0.000000] x86/fpu: Using 'eager' FPU context switches.
[    0.000000] e820: BIOS-provided physical RAM map:
[    0.000000] BIOS-e820: [mem 0x0000000000000000-0x00000000000004ff] reserved
[    0.000000] BIOS-e820: [mem 0x0000000000000500-0x0000000000007bff] usable
[    0.000000] BIOS-e820: [mem 0x0000000000007c00-0x0000000000007dff] reserved
[    0.000000] BIOS-e820: [mem 0x0000000000007e00-0x000000000009fbff] usable
[    0.000000] BIOS-e820: [mem 0x000000000009fc00-0x000000000fffffff] reserved
[    0.000000] BIOS-e820: [mem 0x0000000010000000-0x0000000017ffffff] usable
[    0.000000] BIOS-e820: [mem 0x0000000018000000-0x00000000ffffffff] reserved
[    0.000000] bootconsole [earlyser0] enabled
[    0.000000] Notice: NX (Execute Disable) protection missing in CPU!
[    0.000000] DMI not present or invalid.
[    0.000000] e820: update [mem 0x00000000-0x00000fff] usable ==> reserved
[    0.000000] e820: remove [mem 0x000a0000-0x000fffff] usable
[    0.000000] e820: last_pfn = 0x18000 max_arch_pfn = 0x100000
[    0.000000] MTRR: Disabled
[    0.000000] x86/PAT: MTRRs disabled, skipping PAT initialization too.
[    0.000000] x86/PAT: Configuration [0-7]: WB  WT  UC- UC  WB  WT  UC- UC  
[    0.000000] Scanning 2 areas for low memory corruption
[    0.000000] initial memory mapped: [mem 0x00000000-0x10ffffff]
[    0.000000] Base memory trampoline at [c009b000] 9b000 size 16384
[    0.000000] BRK [0x10d61000, 0x10d61fff] PGTABLE
[    0.000000] RAMDISK: [mem 0x10e43000-0x10fabfff]
[    0.000000] ACPI: Early table checksum verification disabled
[    0.000000] ACPI: RSDP 0x00000000000E0000 000024 (v02 NICTA )
[    0.000000] ACPI: XSDT 0x00000000000E1000 00002C (v01 NICTA  XSDT     00000001 NICT 00000001)
[    0.000000] ACPI: APIC 0x00000000000E102C 000034 (v03 NICTA  APIC     00000003 NICT 00000001)
[    0.000000] ACPI: Local APIC address 0xfee00000
    The faulting Guest OS thread will now be blocked forever.
VM_FATAL_ERROR ::: vmexit handler return error
vmm_print_guest_context:27 | ================== GUEST OS CONTEXT =================
vmm_print_guest_context:30 | exit info : reason 0x30    qualification 0x181   instruction len 0x2 interrupt info 0x0 interrupt error 0x0
vmm_print_guest_context:32 |             guest physical 0xfee00020     rflags 0x210046 
vmm_print_guest_context:34 |             guest interruptibility 0x0   control entry 0x0
vmm_print_guest_context:37 | eip 0xd0040e91
vmm_print_guest_context:39 | eax 0x       0         ebx 0x       0      ecx 0xd0c89fec
vmm_print_guest_context:41 | edx 0xfffff000         esi 0x       0      edi 0x       0
vmm_print_guest_context:43 | ebp 0xd0b19f78
vmm_print_guest_context:45 | cr0 0x80050033      cr3 0x10c86000   cr4 0x2680

--------------------------------------------------------------------------------------------------------

**Comment by WellMcGarnicle** [comment]
Hi @jason416, 
On quick inspection, its not immediately clear to me why your kernel guest is crashing. However we do configure the Linux kernel image for our camkes-vm. These are the configs we use - https://github.com/SEL4PROJ/camkes-vm-linux/tree/master/linux_configs/4.8.16. Are there any major differences between our configuration and yours? 

As a side note, we also have some CMake utility functions that will setup up buildroot and configure it with our given Linux config (See DownloadLinux() and ConfigureLinux() in https://github.com/SEL4PROJ/camkes-vm-linux/blob/master/linux-source-helpers.cmake).

**Comment by jason416** [comment]
thanks for your help, @WellMcGarnicle.  when i see "The Linux guest was built using buildroot, which creates a compatible kernel and minimal root filesystem containing busybox and a in-memory file system (a ramdisk)." , i thought just use buildroot default configuration could do,  but  went wrong.  when i do exercise of adding a module to guest linux, i saw it use the configuration you offered to make a new module and add to rootfs. but what if i want build my own linux kernel, just use 'config' and 'Module.symvers' should work? would you please tell me how to build a linux kernel which can works fine on seL4. really appreciate for your help.

**Comment by WellMcGarnicle** [comment]
No problem, I looked into replicating your issue on my machine by building a custom Linux kernel image. Consequently I've updated the X86 Virtualization page on the doc site with information on how to build your own Linux kernel, found here: https://docs.sel4.systems/VM/CAmkESX86VM#configuring-a-linux-kernel-build. Using the Kbuild settings listed on the doc page should be enough to build a Linux kernel that work in our CAmkES VM.
